privacyidea/lib/tokens/radiustoken.py

```diff
@@ -143,15 +144,193 @@ def update(self, param):
         self.add_tokeninfo("radius.user", val)
         self.add_tokeninfo("tokenkind", TOKENKIND.VIRTUAL)
 
+        log.debug('***[radius token: update]')
+
+    def is_challenge_request(self, passw, user=None, options=None):
+        """
+        This method checks, if this is a request, that triggers a challenge.
+        It depends on the way, the pin is checked - either locally or remote
+
+        :param passw: password, which might be pin or pin+otp
+        :type passw: string
+        :param user: The user from the authentication request
+        :type user: User object
+        :param options: dictionary of additional request parameters
+        :type options: dict
+
+        :return: true or false
+        """
+        otp_counter = -1
+        otpval = passw
+        state = options.get('radius_state')
+
+        log.debug("***[radius token: is challenge request]")
+
+        # should we check the pin locally?
+        if self.check_pin_local:
+            (_res, pin, otpval) = self.split_pin_pass(passw, user,
+                                                      options=options)
+            if not TokenClass.check_pin(self, pin):
+                return False
+
+        # challenge is still valid
+        result = options.get('radius_result')
+        if result is None:
+            res = self.check_radius(passw, options=options, radius_state=state)
+        else:
+            res = result
+
+        return res == 1
+
+    @log_with(log)
+    def create_challenge(self, transactionid=None, options=None):
+        """
+        create a challenge, which is submitted to the user
+
+        :param transactionid: the id of this challenge
+        :param options: the request context parameters / data
+        :return: tuple of (bool, message and data)
+                 bool, if submit was successful
+                 message is submitted to the user
+                 data is preserved in the challenge
+                 attributes - additional attributes, which are displayed in the
+                    output
+        """
+        options = options or {}
+        message = options.get('radius_message') or "Enter your RADIUS tokencode:"
+        state = options.get('radius_state')
+        attributes = {'state': transactionid}
+        validity = int(get_from_config('DefaultChallengeValidityTime', 120))
+
+        log.debug("***[radius token: create challenge]")
```

Comment:
We have a ``log_with`` decorator, so we can remove this line.

Reply:
Done.